README.md
# tfos
TensorFlowOnSpark


## 简介

基于yahoo开源项目TensorFlowOnSpark项目，对TensorFlow和Spark很好的结合，可以实现对大数据进行分布式的机器学习以及深度网络学习。

## 项目结构

```markdown
- deep_insight  平台算子
    - base          平台基础包
    - compile       编译层
    - data          数据层
    - layers        算子层
    - model         模型训练层
    - nets          网络模型层
- docs          文档
- test          网络模型测试
    - test_lenet.py  LeNet网络模型
- tfos          框架打包
    - base          基础模块
    - compile       编译模块
    - layers        模型节点
    - data          数据处理
    - nets          网络模型
    - utils         常用工具方法
    - choices.py    可选参数模块
    - tfos.py       集群框架封装
    - worker.py     集群节点执行模块
```

## 安装

执行以下命令即可安装运行环境：
```
pip install -r requirements.txt
```

[python环境安装](docs/python_env.md)

[TensorFlowOnSpark环境搭建](docs/env.md) 

# 发布

[release](changelog.rst)


changelog.rst
Changelog
=========


0.2.3    ( 2019-09-24 )
-----------------------

New
~~~
- Add alexnet network model. [wwwa]

Fix
~~~
- Debug model train,evaluate,predict. [wwwa]

  - rebuild mnist dataset
  - debug mnist mlp model
  - debug lenet network model

  new: debug evaluate and predict


0.2.3 (2019-09-18)
------------------

New
~~~
- Add recurrent layers and rebuild train model. [wwwa]

  - add recurrent.py
  - add model evaluate and predict

Fix
~~~
- Fix compile optimizer layer. [wwwa]


0.2.1 (2019-09-10)
------------------

New
~~~
- Rebuild neural network struct. [wwwa]

  - add network Model
  - rebuild train and inference operator
  - fix core layer operators
  - fix convolution layer operators
  - fix pooling layer operators
  - add merge layer operator


0.1.1 (2019-09-05)
------------------

New
~~~
- Test train and inference model. [wwwa]
- Add core layers for core.py. [wwwa]
- Add operator annotation. [wwwa]
- Add pooling.py. [wwwa]

  - add MaxPool1D, MaxPool2D, MaxPool3D, AvgPool1D, AvgPool2D, AvgPool3D layer
- Add activation.py. [wwwa]

  - add Activation, LeakyReLU, PReLU, ELU, ThresholdedReLU, Softmax, ReLU
- Add convolution.py. [wwwa]

  - add Convolution1D,Convolution2D,Convolution3D
  - add Convolution2DTranspose, Convolution3DTranspose
  - add SummaryLayer
- Add file_manager.py. [wwwa]
- 增加gitchangelog配置. [wwwa]

Changes
~~~~~~~
- Add mnist cgan mlp operator. [wwwa]

Fix
~~~
- Delete redundancy code. [wwwa]
- Core.py add annotation. [wwwa]
- Cgan network matplot savefig to hdfs. [wwwa]
- Fix operator couldn't execute on deepinsight. [wwwa]

Other
~~~~~
- Test: train model and inference model. [wwwa]
- Docs: add annotation for convolution. [wwwa]
- Add gan network. [wwwa]


0.1.0 (2019-07-26)
------------------

New
~~~
- 增加gitchangelog配置. [wwwa]
- Add .gitchangelog.rc. [wwwa]
- Add test_inference.py. [wwwa]
- Add test_keras_mnist.py. [wwwa]
- Add flatten,convolution,drop,embedding,lstm,max_pooling etc layer.
  [wwwa]
- Change test_train.py. [wwwa]
- Train suanzi. [wwwa]
- Dropout layer and train data type convert. [wwwa]

Changes
~~~~~~~
- Rename examples to deep_insight. [wwwa]
- Package code. [wwwa]
- Fabfile.py add reinstall function. [wwwa]
- Train model result save to h5. [wwwa]
- Test_inference.py. [wwwa]
- Add test_inference.py. [wwwa]
- Add test_inference.py. [wwwa]
- Restore model for model.load_weights() [wwwa]
- Add train model checkpoint. [wwwa]
- Model train relative. [wwwa]
- Add lstm,cnn,multilayer_perceptron. [wwwa]
- Test fabfile. [wwwa]
- Python_env.md. [wwwa]

Fix
~~~
- Add test_optimzer.py and add test_read_mnist.py. [wwwa]
- Test_train.py. [wwwa]

Other
~~~~~
- Debug train and inference. [wwwa]
- Test: add test_keras_model_save. [wwwa]
- Docs: change python_env.md. [wwwa]
- Add fabfile.py and setup.py. [wwwa]
- Add image2tfrecord and perceptron. [wwwa]
- Add read csv file. [wwwa]
- Add tensorflow monitore and checkpoint document. [wwwa]
- Tfos model initialize. [wwwa]
- Build execute graph. [wwwa]
- Packaging TFOS class. [wwwa]
- New regression.py. [wwwa]
- Rebuild project directory. [wwwa]
- Rebuild project directory. [wwwa]
- Modify TensorFlowDatasetAPI.md. [wwwa]
- Modify TensorFlowDatasetAPI.md. [wwwa]
- Add tensorflowonspark架构.md. [wwwa]
- 增加TensorFlowDatasetAPI.md. [wwwa]
- 添加python_env.md文档. [wwwa]
- 修改tensorboard配置路径，添加env.md环境设置文件. [wwwa]
- 增加线性回归代码. [wwwa]
- 将utils目录移动到tfos目录下. [wwwa]
- 添加docs文档目录. [wwwa]
- 去掉对output_data文件夹的跟踪. [wwwa]
- 初始化项目，并实现逻辑归回模型. [wwwa]
- Initial commit. [jinlong]




tfos/graph.py
# -*- coding: utf-8 -*-
"""
:Author  : weijinlong
:Time    : 26/05/2019 19:00
:File    : graph.py
"""

import logging
import os
import time
from collections import namedtuple
from datetime import datetime

import numpy as np
import tensorflow as tf

GRAPH = namedtuple("graph", ['label', 'prediction', 'accuracy', 'loss', 'saver', 'summary_op', 'init_op'])


class Worker(object):

    def __init__(self):
        self.args = None
        self.ctx = None
        self.main_func = None
        self.tf_feed = None
        self.mode = 'train'
        self.model_path = None
        self.model_type = 0  # 0 is classifier, 1 is regression
        self.graph = None

    def get_path(self, type):
        TYPES = ['train', 'inference', 'prediction', 'tensorboard', 'checkpoint']

        if type in TYPES:
            path = os.path.join(self.model_path, type)
            if not os.path.exists(path):
                os.makedirs(path)
        else:
            raise ValueError("model save type is error!")
        return path

    def rdd_data(self):
        tf_feed = self.ctx.get_data_feed(self.mode == 'train')
        tf_feed = self.ctx.get_data_feed(self.mode == 'train')

        def rdd_generator():
            while not tf_feed.should_stop():
                batch = tf_feed.next_batch(1)
                if len(batch) == 0:
                    return
                row = batch[0]
                data = np.array(row[0]).astype(np.float32)
                label = np.array(row[1]).astype(np.int64)
                yield (data, label)

        # Dataset for input data
        ds = tf.data.Dataset.from_generator(rdd_generator, (tf.float32, tf.float32),
                                            (tf.TensorShape([4]), tf.TensorShape([1]))).batch(self.args.batch_size)
        iterator = ds.make_one_shot_iterator()
        x, y_ = iterator.get_next()
        return x, y_

    def tf_data(self):
        pass

    def execute(self):
        worker_num = self.ctx.worker_num
        job_name = self.ctx.job_name
        task_index = self.ctx.task_index

        # Get TF cluster and server instances
        cluster, server = self.ctx.start_cluster_server(1, self.args.rdma)

        if job_name == "ps":
            server.join()
        elif job_name == "worker":
            self.worker(cluster, task_index)
            self.start_session(worker_num, task_index, server.target)

    def worker(self, cluster, task_index):
        with tf.device(
                tf.train.replica_device_setter(worker_device="/job:worker/task:%d" % task_index, cluster=cluster)):
            x, y = self.rdd_data()
            pred, loss, train_op = self.main_func(x, y)
            accuracy = None
            if self.model_type == 0:
                # Test trained model
                label = tf.argmax(y, 1, name="label")
                prediction = tf.argmax(pred, 1, name="prediction")
                correct_prediction = tf.equal(prediction, label)
                accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32), name="accuracy")
                tf.summary.scalar("accuracy", accuracy)

            saver = tf.train.Saver()
            summary_op = tf.summary.merge_all()
            init_op = tf.global_variables_initializer()
            self.graph = GRAPH._make([y, pred, accuracy, loss, saver, summary_op, init_op])

    def start_session(self, worker_num, task_index, master):
        g = self.graph
        # Create a "supervisor", which oversees the training process and stores model state into HDFS
        checkpoint_dir = self.ctx.absolute_path(self.get_path('checkpoint'))
        summary_writer = tf.summary.FileWriter(
            os.path.join(self.get_path('tensorboard'), "tensorboard_%d" % worker_num),
            graph=tf.get_default_graph())

        hooks = [tf.train.StopAtStepHook(last_step=self.args.steps)] if self.mode == "train" else []
        with tf.train.MonitoredTrainingSession(master=master,
                                               is_chief=(task_index == 0),
                                               scaffold=tf.train.Scaffold(init_op=g.init_op,
                                                                          summary_op=g.summary_op,
                                                                          saver=g.saver),
                                               checkpoint_dir=checkpoint_dir,
                                               hooks=hooks) as sess:
            print("{} session ready".format(datetime.now().isoformat()))
            # Loop until the session shuts down or feed has no more data
            step = 0
            while not sess.should_stop() and not self.tf_feed.should_stop():
                if self.mode == "train":
                    _, summary, step = sess.run([self.graph.train_op, g.summary_op, g.global_step])
                    if (step % 2 == 0) and (not sess.should_stop()):
                        # print("{} step: {} accuracy: {}".format(datetime.now().isoformat(), step, sess.run(accuracy)))
                        if self.model_type == 0:
                            logging.error(
                                "{} step: {} accuracy: {}".format(datetime.now().isoformat(), step,
                                                                  sess.run(g.accuracy)))
                        else:
                            logging.error(
                                "{} step: {} loss: {}".format(datetime.now().isoformat(), step, sess.run(g.loss)))
                    if task_index == 0:
                        summary_writer.add_summary(summary, step)
                else:  # args.mode == "inference"
                    labels, preds, acc = sess.run([g.label, g.prediction, g.accuracy])
                    results = ["{} Label: {}, Prediction: {}".format(datetime.now().isoformat(), l, p) for l, p in
                               zip(labels, preds)]
                    self.tf_feed.batch_results(results)

        print("{} stopping MonitoredTrainingSession".format(datetime.now().isoformat()))
        if sess.should_stop() or step >= self.args.steps:
            self.tf_feed.terminate()

            # WORKAROUND FOR https://github.com/tensorflow/tensorflow/issues/21745
            # wait for all other nodes to complete (via done files)
            done_dir = "{}/{}/done".format(self.ctx.absolute_path(self.args.model_path), self.mode)
            print("Writing done file to: {}".format(done_dir))
            tf.gfile.MakeDirs(done_dir)
            with tf.gfile.GFile("{}/{}".format(done_dir, self.ctx.task_index), 'w') as done_file:
                done_file.write("done")

            for i in range(60):
                if len(tf.gfile.ListDirectory(done_dir)) < len(self.ctx.cluster_spec['worker']):
                    print("{} Waiting for other nodes {}".format(datetime.now().isoformat(), i))
                    time.sleep(1)
                else:
                    print("{} All nodes done".format(datetime.now().isoformat()))
                    break

    def __call__(self, args, ctx):
        self.args = args
        self.ctx = ctx
        self.execute()
